File: linda-core/src/main/java/it/unibo/coordination/utils/events/SyncEventEmitter.java
package it.unibo.coordination.utils.events;

public interface SyncEventEmitter<Arg> {

    static <X> SyncEventEmitter<X> ordered() {
        return new SyncOrderedEventSourceImpl<>();
    }

    Arg syncEmit(Arg data);

    EventSource<Arg> getEventSource();

}

File: utils/src/main/kotlin/it/unibo/coordination/utils/events/SyncEventEmitter.kt
package it.unibo.coordination.utils.events

interface SyncEventEmitter<T> : EventEmitter<T> {

    fun syncEmit(event: T): T

    @JvmDefault
    override fun emit(event: T) {
        syncEmit(event)
    }

    companion object {

        @JvmStatic
        fun <X> ordered(): SyncEventEmitter<X> {
            return SyncOrderedEventSource()
        }
    }

}